Import and refactor LingoX code from the Edraak repo

OmarIthawi opened the pull request:
Closes #7.

### How to Test
 - `vagrant ssh` and activate the `edxapp` environment
 - `$ pip install -e git+https://github.com/appsembler/LingoX.git@omar/migrate#egg=lingox`
 - Add `lingox` to the `ADDL_INSTALLED_APPS`.
 - Set `LANGUAGE_CODE` to `ar`
 - Reload the LMS server
 - Open a new incognito window on `http://localhost:8000/`, you should see an Arabic interface

I'm planning to do minor improvements including docs and Sites (aka microsites) support. More on the issues:

 - https://github.com/appsembler/LingoX/issues

OmarIthawi committed:
Import and refactor LingoX code from the Edraak repo

OmarIthawi committed:
Added docs

OmarIthawi committed:
Merge pull request #8 from appsembler/omar/docs

Added docs

OmarIthawi committed:
Support Django "Sites" and edX site_configuration

OmarIthawi committed:
Merge pull request #9 from appsembler/omar/sites-support

Support Django "Sites" and edX site_configuration

OmarIthawi committed:
Configurable API prefixes

OmarIthawi committed:
Merge pull request #10 from appsembler/omar/api-endpoints-configs

Configurable API prefixes

OmarIthawi commented:
@amirtds @bryanlandia this PR is ready for review. Appreciate it if you can take a look here.

OmarIthawi committed:
Fix git URL

bryanlandia commented:
@OmarIthawi Does this target any specific edX release?

OmarIthawi commented:
@bryanlandia it's mainly for Ficus. Sorry didn't see your comment earlier.

bryanlandia commented on README.rst:
```diff
@@ -1,49 +1,96 @@
 
-Overview (please modify)
-------------------------
+The edX Platform and other Django-based is bundled by default with ``LocaleMiddleware``
```
"other Django-based ...?"

bryanlandia commented on README.rst:
```diff
@@ -1,49 +1,96 @@
--------------
+This package helps to enforce ``settings.LANGUAGE_CODE`` (or ``site_configuration`` provided
+value) as the site's default language for the edX Platform.
+This package do allow learners to change the language. This package
```
s/do/does

bryanlandia commented on README.rst:
```diff
@@ -1,49 +1,96 @@
+This package helps to enforce ``settings.LANGUAGE_CODE`` (or ``site_configuration`` provided
+value) as the site's default language for the edX Platform.
+This package do allow learners to change the language. This package
+also maintains backward compatibility with API endpoints and pass the ``Accept-Language``
```
s/pass/passes

bryanlandia commented on README.rst:
```diff
@@ -1,49 +1,96 @@
+- ``$ pip install -e git+https://github.com/appsembler/LingoX.git#egg=lingox``
+- Add ``lingox`` to the ``ADDL_INSTALLED_APPS`` in the ``lms.env.json`` (or your ``server-vars.yml``)
+- Set ``LANGUAGE_CODE`` to ``ar``
+- To configure a different language for a specific site go to:
```
did you mean to leave this incomplete after "for a specific site go to"?

bryanlandia commented on README.rst:
```diff
@@ -1,49 +1,96 @@
+---------------
+This module monkey-patches the edX platform the following way:
+
+- Add the ``DefaultLocaleMiddleware`` to ``MIDDLEWARE_CLASSES`` before any *known* locale-aware middleware
```
s/middleware/middleware. (add period)

bryanlandia commented on lingox/apps.py:
```diff
@@ -14,3 +17,9 @@ class LingoXConfig(AppConfig):
+        """
+        Monkeypatch MIDDLEWARE_CLASSES to the LingoX middleware.
+        """
+        settings.MIDDLEWARE_CLASSES = add_locale_middleware(settings.MIDDLEWARE_CLASSES)
```
I like to store the original value before monkey-patching, like

```
MIDDLEWARE_CLASSES_ORIG = settings.MIDDLEWARE_CLASSES
```
and preferably stored at the module level, not within function scope.  
This is just a precaution in case something goes wrong, at least it gives you easier access if you have to access in a running process like with gdb or the like.

OmarIthawi replied:
It's being logged by `add_locale_middleware`, so I suppose that's sufficient.

bryanlandia commented on lingox/middleware.py:
```diff
@@ -0,0 +1,39 @@
+        """
+        Change the request's `HTTP_ACCEPT_LANGUAGE` to `settings.LANGUAGE_CODE`.
+        """
+        # This middleware is only needed for regular browser pages, it's effect is breaking the behaviour on the
```
s/it's/its

bryanlandia commented on lingox/helpers.py:
```diff
@@ -0,0 +1,88 @@
+        '/api/',
+        '/user_api/',
+        '/notifier_api/',
+    ]
```
I wonder if there's a way to get at these programmatically.

OmarIthawi replied:
I wanted to have a way, but couldn't find a proper method.

bryanlandia commented on tests/edx_platform_mock/test_site/urls.py:
```diff
@@ -0,0 +1,19 @@
+urlpatterns = [
+    url(r'^api/$', views.api),
+    url(r'^user_api/$', views.api),
+    url(r'^notifier_api/$', views.api),
```
plus `discussion_api`?

bryanlandia commented on tests/tests.py:
```diff
@@ -0,0 +1,307 @@
+        {'path': '/', 'should_be_api': False},
+        {'path': '/user_api/', 'should_be_api': True},
+        {'path': '/notifier_api/', 'should_be_api': True},
+        {'path': '/reporting/api/', 'should_be_api': False},  # By default, LingoX don't know about reporting
```
plus `'/discussions_api/`

bryanlandia commented on README.rst:
```diff
@@ -1,49 +1,96 @@
+- Set ``LANGUAGE_CODE`` to ``ar``
+- To configure a different language for a specific site go to:
+- Reload the server
+- Open a new incognito window on ``http://localhost:8000/``, you should see an Arabic interface
```
You would need to have pulled down the translations, as well.

OmarIthawi replied:
edX platform is always bundled with Arabic, among other languages. Unless something's broken, usually it works on both `open-release/ficus.master` and `appsembler/ficus/master` without modification.

However, the Arabic translation is no 100% complete, therefore there will be English words here and there.

bryanlandia reviewed (changes requested):
Very cool, @OmarIthawi !  It worked for me after I pulled down 'ar' translations and ran `paver i18n_fastgenerate` and disabled the dark_lang config.  I wasn't able to run the tests, I got this error:

```
edxapp@vagrant:~/venvs/edxapp/src/lingox$ DJANGO_SETTINGS_MODULE=test_settings pytest
========================================================= test session starts =========================================================
platform linux2 -- Python 2.7.12, pytest-3.2.3, py-1.4.34, pluggy-0.4.0
rootdir: /edx/app/edxapp/venvs/edxapp/src/lingox, inifile: tox.ini
plugins: cov-2.5.1, flaky-3.3.0
collected 0 items / 1 errors                                                                                                           


---------- coverage: platform linux2, python 2.7.12-final-0 ----------
Name                   Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------------------
lingox/__init__.py         3      0      0      0   100%
lingox/apps.py             9      9      0      0     0%   2-25
lingox/helpers.py         27     20     10      0    19%   23-61, 76-88
lingox/middleware.py      11      7      4      0    27%   12-39
------------------------------------------------------------------
TOTAL                     50     36     14      0    22%
Coverage XML written to file coverage.xml

===Flaky Test Report===


===End Flaky Test Report===
=============================================================== ERRORS ================================================================
___________________________________________________ ERROR collecting tests/tests.py ___________________________________________________
ImportError while importing test module '/edx/app/edxapp/venvs/edxapp/src/lingox/tests/tests.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/tests.py:14: in <module>
    from lingox.middleware import DefaultLocaleMiddleware
lingox/middleware.py:9: in <module>
    from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
../../../../edx-platform/openedx/core/djangoapps/site_configuration/helpers.py:6: in <module>
    from microsite_configuration import microsite
E   ImportError: No module named microsite_configuration
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

OmarIthawi committed:
Addressing Bryan's review comments

OmarIthawi commented:
@bryanlandia I've addressed all of your comments. Regarding the tests, you need a different python virtual environment (not `edxapp_env`). I've added instructions in the "How to Develop" section in README.

bryanlandia commented on README.rst:
```diff
@@ -69,11 +71,32 @@ Monkey Patching
+   $ mkvirtualenv lingox
+   $ make requirements
+   $ make quality
+   $ make test
```
```
$ mkvirtualenv lingox
...
$ (lingox) make requirements
...
$ (lingox) make quality
...
$ (lingox) make test
...
ImportError: No module named test_site
make: *** [test] Error 1
```

`test` using these instructions doesn't work for me.

bryanlandia commented on tests/edx_platform_mock/test_site/urls.py:
```diff
@@ -11,6 +11,7 @@
     url(r'^api/$', views.api),
     url(r'^user_api/$', views.api),
     url(r'^notifier_api/$', views.api),
+    url(r'^api/discussion/$', views.api),
```
Sorry, Omar, I thought there was a `discussion_api` url prefix.  If not, you don't really need this.  as it's redundent with plain `api/`.

OmarIthawi replied:
![image](https://user-images.githubusercontent.com/645156/41233486-4090e12c-6d92-11e8-9a6f-dce07f4b186a.png)

I'll remove it then.

bryanlandia reviewed (changes requested):
Looks good except for still having trouble running the tests following the instructions.

OmarIthawi committed:
Fix test_site bug

OmarIthawi commented:
@bryanlandia the `test_site` bug has been fixed. Please run `make requirements` again, it should work:

```
$ git fetch
$ git checkout omar/migrate
$ git reset --hard origin/omar/migrate
$ workon lingox
$ make requirements
```

OmarIthawi commented:
Thanks @bryanlandia! No worries about the delay!
